### PR TITLE
Add support for missing in ScriptSortBuilder

### LINKED ISSUE #34597: Add support for missing in ScriptSortBuilder
`FieldSortBuilder` allows handling of null values through `missing` (`_first` and `_last`). 
As scripts can return `null`, the same feature would be quite convenient to have in `ScriptSortBuilder` since otherwise, scripts always have to return a non-`null` value which is tricky to position (first vs last).

### PR DESCRIPTION
**Pull request by affanv14**
Closes #34597

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java [hunk 1]
```diff
@@ -124,6 +128,7 @@ public ScriptSortBuilder(StreamInput in) throws IOException {
         if (in.getVersion().onOrAfter(Version.V_6_1_0)) {
             nestedSort = in.readOptionalWriteable(NestedSortBuilder::new);
         }
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java]: You need to check the version of the input and read the value only if it is on the current version. See above for an example of an option added in v6.1.0.

File: server/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java [hunk 2]
```diff
@@ -137,6 +142,7 @@ public void writeTo(StreamOutput out) throws IOException {
         if (out.getVersion().onOrAfter(Version.V_6_1_0)) {
             out.writeOptionalWriteable(nestedSort);
         }
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java]: Same here:
`if (out.getVersion().onOrAfter(Version.V7_.0_0))...`

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @affanv14 , this is a good start. However your change is a not implementing the requested feature. The idea is to be able to return `null` in a sort script. Currently returning null in a script that returns string is translated into the string "null" and returning null in a script that returns number is not allowed. If we want to allow null values as a valid return type you'll need to change the anonymous `BytesRefFieldComparatorSource` in `ScriptSortBuilder` to something like:
```
fieldComparatorSource = new BytesRefFieldComparatorSource(null, "_first", valueMode, nested) {
                    StringSortScript leafScript;
                    @Override
                    protected SortedBinaryDocValues getValues(LeafReaderContext context) throws IOException {
                        leafScript = searchScript.newInstance(context);
                        final BinaryDocValues values = new AbstractBinaryDocValues() {
                            final BytesRefBuilder spare = new BytesRefBuilder();
                            @Override
                            public boolean advanceExact(int doc) throws IOException {
                                leafScript.setDocument(doc);
                                String ret = leafScript.execute();
                                if (ret == null) {
                                    return false;
                                }
                                spare.copyChars(leafScript.execute());
                                return true;
                            }
                            @Override
                            public BytesRef binaryValue() {
                                return spare.get();
                            }
                        };
                        return FieldData.singleton(values);
                    }
                    @Override
                    protected void setScorer(Scorable scorer) {
                        leafScript.setScorer(scorer);
                    }
                };
```
And the `DoubleValuesComparatorSource` to:
```
fieldComparatorSource = new DoubleValuesComparatorSource(null, Double.MAX_VALUE, valueMode, nested) {
                    NumberSortScript leafScript;
                    @Override
                    protected SortedNumericDoubleValues getValues(LeafReaderContext context) throws IOException {
                        leafScript = numberSortScript.newInstance(context);
                        final NumericDoubleValues values = new NumericDoubleValues() {
                            double value;

                            @Override
                            public boolean advanceExact(int doc) throws IOException {
                                leafScript.setDocument(doc);
                                Double ret = leafScript.execute();
                                if (ret == null) {
                                    return false;
                                }
                                value = ret;
                                return true;
                            }

                            @Override
                            public double doubleValue() {
                                return value;
                            }
                        };
                        return FieldData.singleton(values);
                    }
                    @Override
                    protected void setScorer(Scorable scorer) {
                        leafScript.setScorer(scorer);
                    }
                };
```
However `NumberSortScript` returns a `double` currently so you'll also need to change it to return a `Double` instead.
Finally we need tests that check that `null` values are correctly handled. This is the only way to make sure that the change works ;).

**Comment by affanv14** [comment]
Sorry it took so long, life got in the way :)

**Comment by affanv14** [comment]
I still have a few more changes to do though

**Comment by affanv14** [comment]
Alright @jimczi, I finished making the changes

**Review by jimczi — COMMENTED** [review]
Thanks @affanv14, the change looks good to me but I'd like someone more familiar with scripts to review as well which is why I asked @rjernst to take a look when he has time.

**Comment by rjernst** [comment]
I'm sorry it has taken me so long to look at this. Thank you for the work here, but I don't think we should do this. Changing the return signature of double sort scripts from `double` to `Double` has a performance penalty due to boxing. It actually took us a large amount of work to finally get to using non-boxed types, made possible by script contexts.

When using scripts for sorting, the user is taking on the responsibility to determine a value for every document. Adding some external logic to that is unnecessary; the script can determine what to do when a field is missing. Additionally, expressions already have logic for missing fields that is hardcoded (eg 0.0d for numeric fields). Thus expressions would never return null, and the behavior would be inconsistent between painless and expressions (while this could be changed, it would be a large breaking change).

**Comment by affanv14** [comment]
Thanks @rjernst for the review! 

So I guess support for missing in ScriptSortBuilder isn't needed. Shall I go ahead and close the pull request?

**Pull request closed without merging**